### PR TITLE
Adding ADAL wrapper to msrestazure

### PR DESCRIPTION
**Pull request by lmazuel**
After discussion with @begoldsm, this allows people other than us (Python team or CLI team) to connect using ADAL, without forcing on ADAL features. 

Samples:
```python
from msrestazure.azure_active_directory import AdalAuthentication
import adal

context = adal.AuthenticationContext('https://login.microsoftonline.com/common')
token_retriver = lambda: context.acquire_token_with_username_password(
    'https://management.core.windows.net/',
    'unittest@myaccountoutlook.onmicrosoft.com',
    'PASSWORD!!!',
    '04b07795-8ddb-461a-bbee-02f9e1bf7b46' # Xplat CLI or yours
)
return AdalAuthentication(token_retriver)
```

This is a 99% copy/paste from CLI code, I almost didn't code anything here, just backport the code from @yugangw-msft. I didn't include ADAL in `install_requires` for now on purpose, I think it's more interesting to do it in `extra_requires`. And be clear in the documentation. Also the main advantage here is that we don't own any ADAL features, so any new stuff can be used directly without changing `msrestazure`.

@annatisch @brettcannon what do you think?

### CODE REVIEW

**Comment by lmazuel** [comment]
@brettcannon also suggests:

```python
from msrestazure.azure_active_directory import AdalAuthentication
import adal

context = adal.AuthenticationContext('https://login.microsoftonline.com/common')
return AdalAuthentication(
    context.acquire_token_with_username_password
    'https://management.core.windows.net/',
    'unittest@myaccountoutlook.onmicrosoft.com',
    'PASSWORD!!!',
    '04b07795-8ddb-461a-bbee-02f9e1bf7b46' # Xplat CLI or yours
)
```

IOW, use args/kwargs to avoid the lambda syntax

**Comment by lmazuel** [comment]
Adding support for @brettcannon syntax. Now both syntax (lambda + args/kwargs) are supported the same, depending of the coders habits.

File: msrestazure/azure_active_directory.py [hunk 1]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
...
+        import adal # Adal is not mandatory
+
+        try:
+            raw_token = self._adal_method(*self._args, **self._kwargs)
+            scheme, token = raw_token['tokenType'], raw_token['accessToken']
+        except adal.AdalError as err:
+            #pylint: disable=no-member
+            if (hasattr(err, 'error_response') and ('error_description' in err.error_response)
+                    and ('AADSTS70008:' in err.error_response['error_description'])):
+                raise Expired("Credentials have expired due to inactivity. Please run 'az login'")
```
Comment (begoldsm) [inline msrestazure/azure_active_directory.py]: consider changing this error message, since this is not the CLI and they can't run az login.

**Comment by begoldsm** [comment]
:shipit:

File: msrestazure/azure_active_directory.py [hunk 2]
```diff
@@ -39,10 +39,16 @@
     MismatchingStateError,
     OAuth2Error,
     TokenExpiredError)
-from requests import RequestException
+from requests import (
+    RequestException,
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: There's no need to make this span multiple lines. Same goes for other imports.

File: msrestazure/azure_active_directory.py [hunk 3]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
             raise_with_traceback(AuthenticationError, "", err)
         else:
             self.token = token
+
+class AdalAuthentication(Authentication):#pylint: disable=too-few-public-methods
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: Two spaces between `:` and `#`.

Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: Probably a docstring that explains what the class is about. Same goes for methods.

File: msrestazure/azure_active_directory.py [hunk 4]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
...
+
+    def __init__(self, adal_method, *args, **kwargs):
+        self._adal_method = adal_method
+        self._args = args
+        self._kwargs = kwargs
+
+    def signed_session(self):
+        session = super(AdalAuthentication, self).signed_session()
+
+        import adal # Adal is not mandatory
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: If you make adal mandatory then this should move to the top of the file.

File: msrestazure/azure_active_directory.py [hunk 5]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
...
+
+        try:
+            raw_token = self._adal_method(*self._args, **self._kwargs)
+            scheme, token = raw_token['tokenType'], raw_token['accessToken']
+        except adal.AdalError as err:
+            #pylint: disable=no-member
+            if (hasattr(err, 'error_response') and ('error_description' in err.error_response)
+                    and ('AADSTS70008:' in err.error_response['error_description'])):
+                raise Expired("Credentials have expired due to inactivity.")
+
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: Replace the blank line with an `else:` clause.

File: msrestazure/azure_active_directory.py [hunk 6]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
...
+            raise AuthenticationError(err)
+        except ConnectionError as err:
+            raise AuthenticationError('Please ensure you have network connection. Error detail: ' + str(err))
+
+        header = "{} {}".format(scheme, token)
+        session.headers['Authorization'] = header
+        return session
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: Missing a newline at the end of the file.

File: msrestazure/azure_active_directory.py [hunk 7]
```diff
@@ -525,3 +531,32 @@ def set_token(self, response_url):
...
+        self._kwargs = kwargs
+
+    def signed_session(self):
+        session = super(AdalAuthentication, self).signed_session()
+
+        import adal # Adal is not mandatory
+
+        try:
+            raw_token = self._adal_method(*self._args, **self._kwargs)
+            scheme, token = raw_token['tokenType'], raw_token['accessToken']
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: This should be moved outside of the `try` block as it won't trigger an `AdalError`.

**Comment by lmazuel** [comment]
@brettcannon I think I made the changes
@annatisch nothing to say?

**Comment by annatisch** [comment]
@lmazuel  - not really, it's nice and simple :).
Unless you want to hook in the auto-token-refreshing?

File: test/unittest_auth.py [hunk 8]
```diff
@@ -356,6 +359,34 @@ def test_user_pass_credentials(self):
...
+            session = credentials.signed_session()
+
+        def connection_error():
+            raise ConnectionError("Plug the network")
+        credentials = AdalAuthentication(connection_error)
+        with self.assertRaises(AuthenticationError) as cm:
+            session = credentials.signed_session()
 
 if __name__ == '__main__':
     unittest.main()
```
Comment (brettcannon) [inline test/unittest_auth.py]: Missing a newline at the end of the file.

File: msrestazure/azure_active_directory.py [hunk 9]
```diff
@@ -525,3 +524,81 @@ def set_token(self, response_url):
             raise_with_traceback(AuthenticationError, "", err)
         else:
             self.token = token
+
+
+class AdalAuthentication(Authentication):  # pylint: disable=too-few-public-methods
+    """A wrapper to use ADAL for Python easily to authenticate on Azure.
```
Comment (brettcannon) [inline msrestazure/azure_active_directory.py]: Put the closing `"""` on the same line and have a newline between it and `__init__()`.

File: test/unittest_auth.py (hunk 8 above)
Comment (lmazuel) [inline test/unittest_auth.py]: done this one

File: msrestazure/azure_active_directory.py (hunk 9 above)
Comment (lmazuel) [inline msrestazure/azure_active_directory.py]: done too

Comment (lmazuel) [inline msrestazure/azure_active_directory.py]: General question: should I PEP8 or/and flake8 or/and PyLint or/and mypy?

**Pull request merged**